hector_math/include/hector_math/containers/ring_buffer.h

```diff
@@ -0,0 +1,270 @@
+//
+// Created by aljoscha on 02.02.2023.
+//
+
+#ifndef HECTOR_MATH_RING_BUFFER_H
+#define HECTOR_MATH_RING_BUFFER_H
+
+#include <array>
+#include <assert.h>
+#include <type_traits>
+
+namespace hector_math
+{
+
+template<typename T, int MaxSize>
+class RingBuffer
+{
+public:
+  // template parameter controls whether the iterator returns const references
+  template<bool IS_CONST>
+  struct ring_iterator {
+    using iterator_category = std::random_access_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    // using difference_type = typename std::iterator<std::random_access_iterator_tag, T>::difference_type;
+    using value_type = T;
+    using pointer = T *;
+    using reference = T &;
+    using const_reference = T const &;
+
+    ring_iterator( RingBuffer<T, MaxSize> *buffer, size_t index )
+        : buffer_( buffer ), index_( index )
+
+    {
+    }
+
+    template<bool Z = IS_CONST, typename std::enable_if<( !Z ), int>::type * = nullptr>
+    reference operator*() noexcept
+    {
+      return ( *buffer_ )[index_];
+    }
+    const_reference operator*() const noexcept { return ( *buffer_ )[index_]; }
+    template<bool Z = IS_CONST, typename std::enable_if<( !Z ), int>::type * = nullptr>
+    reference operator->() noexcept
+    {
+      return &( ( *buffer_ )[index_] );
+    }
+    const_reference operator->() const noexcept { return &( ( *buffer_ )[index_] ); }
+
+    ring_iterator<IS_CONST> &operator++()
+    {
+      index_++;
+      if ( index_ > MaxSize )
+        index_ = 0;
+      return *this;
+    }
+
+    ring_iterator<IS_CONST> &operator--()
+    {
+      index_ = ( index_ + MaxSize ) % ( MaxSize + 1 ); // index-- and back to 0 if > MaxSize
```

Comment:
I'd probably rename `MaxSize` to `MaxElements` and add a constexpr member `StorageSize` with the +1 to reduce the risk of forgetting the +1 somewhere.